### PR TITLE
ref(catalog): remove GetMeshSpec from MeshCataloger

### LINKED ISSUE #3148: Remove GetSMISpec() from MeshCataloger interface
This GitHub Issue is to change [the **MeshCataloger** interface](https://github.com/openservicemesh/osm/blob/release-v0.8/pkg/catalog/types.go#L62-L132) and remove [GetSMISpec()](https://github.com/openservicemesh/osm/blob/release-v0.8/pkg/catalog/types.go#L64)

The `GetSMISpec()` is confusing at best. This exposes all of SMI and is used in `pkg/envoy/lds/inmesh.go` only to `ListTrafficSplits`: https://github.com/openservicemesh/osm/blob/1f10701/pkg/envoy/lds/inmesh.go#L355

But we only need to ListTrafficSplits in order to:
1. create a `[]*xds_tcp_proxy.TcpProxy_WeightedCluster_ClusterWeight` [here](https://github.com/openservicemesh/osm/blob/1f10701/pkg/envoy/lds/inmesh.go#L352)
2. create a set of services `map[service.MeshService]struct{}` [here](https://github.com/openservicemesh/osm/blob/1f10701/pkg/envoy/lds/inmesh.go#L420)

For 1 I propose we add this to `MeshCataloger` interface:
```go
// GetWeightedClustersForUpstream lists the apex services from traffic split policies for a given upstream
GetWeightedClustersForUpstream(upstream service.MeshService) []*xds_tcp_proxy.TcpProxy_WeightedCluster_ClusterWeight
```

For 2 I propose we add this to `MeshCataloger` interface:
```go
// ListMeshServiceForServiceAccount lists the services for a given service account
ListMeshServiceForServiceAccount(service.K8sServiceAccount) map[service.MeshService]struct{}
```

and finally we can remove `GetSMISpec`

---

I propose we remove it.

To do that we could:
  1. Add  `ListMeshServiceForServiceAccount` 
  2. Add `GetWeightedClustersForUpstream`
  3. Remove [GetSMISpec()](https://github.com/openservicemesh/osm/blob/release-v0.8/pkg/catalog/types.go#L64)

---

I started trying to unravel this in this PR, **which should be eventually deleted** - this was just o POC to illustrate a **possible** refactor:  `[RFC] catalog: Removing GetSMISpec; Adding ListMeshServiceForServiceAccount and GetWeightedClustersForUpstream to MeshCatalog interface` #3110

### PR DESCRIPTION
**Pull request by nojnhuh**
**Description**:
This change removes the `GetMeshSpec()` method from the `MeshCataloger`
interface. The necessary functionality has been moved into two new
methods on the `MeshCataloger` interface:

    // GetWeightedClustersForUpstream lists the apex services from traffic split policies for a given upstream.
    GetWeightedClustersForUpstream(service.MeshService) []*xds_tcp_proxy.TcpProxy_WeightedCluster_ClusterWeight

    // ListMeshServicesForIdentity lists the services for a given service identity.
    ListMeshServicesForIdentity(identity.ServiceIdentity) []service.MeshService

No functional changes.

Largely based off of @draychev's work in #3110

Fixes #3148
Closes #3110

**Affected area**:

- New Functionality      [ ]
- Documentation          [ ]
- Install                [ ]
- Control Plane          [X]
- CLI Tool               [ ]
- Certificate Management [ ]
- Networking             [ ]
- Metrics                [ ]
- SMI Policy             [ ]
- Security               [ ]
- Tests                  [ ]
- CI System              [ ]
- Demo                   [ ]
- Performance            [ ]
- Other                  [ ]

Please answer the following questions with yes/no.

- Does this change contain code from or inspired by another project? If so, did you notify the maintainers and provide attribution?
No

### CODE REVIEW

File: pkg/catalog/outbound_traffic_policies_test.go [hunk 1]
```diff
@@ -996,3 +997,283 @@ func TestGetDestinationServicesFromTrafficTarget(t *testing.T) {
...
+
+			clusterWeights := mc.GetWeightedClustersForUpstream(tc.upstream)
+
+			assert := tassert.New(t)
+			assert.Equal(tc.expected, clusterWeights)
+		})
+	}
+}
+
+func TestListMeshServicesForIdentity(t *testing.T) {
```
Comment (nojnhuh) [inline pkg/catalog/outbound_traffic_policies_test.go:1184]: This test is new, not based off any other existing tests.

**Review by draychev — APPROVED** [review]
This looks fantastic to me!
Hoping e2es pass too 🤞🏻

File: pkg/catalog/outbound_traffic_policies.go [hunk 2]
```diff
@@ -201,3 +202,92 @@ func (mc *MeshCatalog) getDestinationServicesFromTrafficTarget(t *access.Traffic
 	}
 	return destServices, nil
 }
+
+// GetWeightedClustersForUpstream returns Envoy cluster weights for the given
+// upstream service, the apex service of a TrafficSplit.
+func (mc *MeshCatalog) GetWeightedClustersForUpstream(upstream service.MeshService) []*xds_tcp_proxy.TcpProxy_WeightedCluster_ClusterWeight {
```
Comment (shashankram) [inline pkg/catalog/outbound_traffic_policies.go:208]: The return type here seems weird to me. The function returns a weighted cluster for an upstream regardless of its protocol, but we are returning a XDS TcpProxy object here. Could you use `service.WeightedCluster` instead?

We have been intentional in not exposing XDS structs in the catalog interface so far. Has this changed?

**Review by shashankram — CHANGES_REQUESTED** [review]
Code looks good overall, except the fact that we are exposing an XDS TcpProxy object in a generic interface seems off to me.

File: pkg/catalog/outbound_traffic_policies.go (hunk 2 above)
Comment (nojnhuh) [reply pkg/catalog/outbound_traffic_policies.go]: Makes sense. Just fixed.

File: pkg/catalog/types.go [hunk 3]
```diff
@@ -93,6 +90,12 @@ type MeshCataloger interface {
 
 	// ListInboundTrafficTargetsWithRoutes returns a list traffic target objects composed of its routes for the given destination service identity
 	ListInboundTrafficTargetsWithRoutes(identity.ServiceIdentity) ([]trafficpolicy.TrafficTargetWithRoutes, error)
+
+	// GetWeightedClustersForUpstream lists the apex services from traffic split policies for a given upstream.
```
Comment (shashankram) [inline pkg/catalog/types.go:94]: This comment needs to be updated to reflect what is being returned, ie. weighted cluster backends corresponding to the upstream service.

**Review by shashankram — APPROVED** [review]

File: pkg/catalog/types.go (hunk 3 above)
Comment (nojnhuh) [reply pkg/catalog/types.go]: Thanks, fixed.

**Pull request merged**